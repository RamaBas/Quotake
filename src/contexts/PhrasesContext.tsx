import React, { createContext, useState, useCallback, useContext } from 'react';
import type { Phrase, PhrasesContextType } from '../types/phrases';

const PhrasesContext = createContext<PhrasesContextType | undefined>(undefined);

export const PhrasesProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {

  // Falta desarrollar addPhrase para dejr de mockear esto
  const [phrases, setPhrases] = useState<Phrase[]>([
    { id: '1', text: 'Aquel que a buen árbol se arrima, buena sombra recibe' },
    { id: '2', text: 'La vida es lo que pasa mientras estás ocupado haciendo Challanges.' },
    { id: '3', text: 'La simplicidad es la máxima sofisticación.' },
    { id: '4', text: 'El único modo de hacer un gran trabajo es amar lo que haces.' },
  ]);

  const addPhrase = useCallback((text: string) => {
    setPhrases(prev => [
      ...prev,
      {
        id: Date.now().toString(),
        text,
      }
    ]);
  }, []);

  const deletePhrase = useCallback((id: string) => {
    setPhrases(prev => prev.filter(phrase => phrase.id !== id));
  }, []);

  const searchPhrases = useCallback((text: string) => {
    return text 
      ? phrases.filter(phrase => 
          phrase.text.toLowerCase().includes(text.toLowerCase())
        )
      : phrases;
  }, [phrases]);

  return (
    <PhrasesContext.Provider 
      value={{ phrases, addPhrase, deletePhrase, searchPhrases }}
    >
      {children}
    </PhrasesContext.Provider>
  );
};

export const usePhrases = (): PhrasesContextType => {
  const context = useContext(PhrasesContext);
  if (!context) {
    throw new Error('error en usePhrases');
  }
  return context;
};