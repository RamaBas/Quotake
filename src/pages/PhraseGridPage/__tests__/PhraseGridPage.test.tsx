import { render, screen, fireEvent } from '@testing-library/react';
import PhrasesGridPage from '../PhraseGridPage';
import { usePhrases } from '../../../contexts/PhrasesContext';

interface SearchBarProps {
    searchTerm: string;
    onSearchChange: (term: string) => void;
}
  
interface PhrasesGridWrapperProps {
    phrases: Array<{ id: string; text: string }>;
    onDelete?: (id: string) => void;
    emptyMessage?: string;
}
  
interface DeleteModalProps {
    open: boolean;
    onClose: () => void;
    onConfirm: () => void;
}

jest.mock('../../../components/SearchBar/SearchBar', () => ({
    __esModule: true,
    default: ({ onSearchChange }: SearchBarProps) => (
        <input
        data-testid="search-bar-mock"
        type="text"
        onChange={(e) => onSearchChange(e.target.value)}
        />
    )
}));

jest.mock('../../../components/PhrasesGridWrapper/PhrasesGridWrapper', () => ({
    __esModule: true,
    default: ({ onDelete }: PhrasesGridWrapperProps) => (
        <div 
        data-testid="phrases-grid-mock"
        onClick={() => onDelete && onDelete('1')}
        />
    )
}));

jest.mock('../../../components/DeleteModal/DeleteModal', () => ({
    __esModule: true,
    default: ({ open, onConfirm }: DeleteModalProps) => (
        <div 
        data-testid="delete-modal-mock"
        data-open={open ? 'true' : 'false'}
        onClick={() => onConfirm()}
        />
    )
}));

jest.mock('../../../contexts/PhrasesContext');

interface MockPhrase {
    id: string;
    text: string;
}

interface MockUsePhrases {
    phrases: MockPhrase[];
    deletePhrase: jest.Mock;
    searchPhrases: (term: string) => MockPhrase[];
    addPhrase: jest.Mock;
}
  
const mockUsePhrases: MockUsePhrases = {
    phrases: [
        { id: '1', text: 'Frase de prueba 1' },
        { id: '2', text: 'Frase de prueba 2' }
    ],
    deletePhrase: jest.fn(),
    searchPhrases: jest.fn((term: string) => 
        mockUsePhrases.phrases.filter(p => p.text.includes(term))
    ),
    addPhrase: jest.fn() // Añadido para cumplir con la interfaz
};

beforeEach(() => {
    (usePhrases as jest.Mock).mockReturnValue(mockUsePhrases);
});

describe('PhrasesGridPage', () => {
    it('renders correctly with phrases', () => {
        const { asFragment } = render(<PhrasesGridPage />);

        expect(screen.getByTestId('phrases-grid-mock')).toBeInTheDocument();
        expect(screen.getByTestId('search-bar-mock')).toBeInTheDocument();

        expect(asFragment()).toMatchSnapshot();
    });

    it('renders empty state when no phrases', () => {
        (usePhrases as jest.Mock).mockReturnValueOnce({
        ...mockUsePhrases,
        phrases: [],
        searchPhrases: jest.fn(() => [])
        });
    
        render(<PhrasesGridPage />);
        
        expect(screen.getByTestId('phrases-grid-mock')).toBeInTheDocument();
        expect(screen.queryByTestId('search-bar-mock')).not.toBeInTheDocument();
    });

    it('handles delete flow correctly', () => {
        render(<PhrasesGridPage />);
        
        // Simular click en delete
        fireEvent.click(screen.getByTestId('phrases-grid-mock'));
        
        // Verificar que el modal se abre
        const deleteModal = screen.getByTestId('delete-modal-mock');
        expect(deleteModal).toHaveAttribute('data-open', 'true');
        
        // Simular confirmación
        fireEvent.click(deleteModal);
        
        expect(mockUsePhrases.deletePhrase).toHaveBeenCalledWith('1');
    });

    it('Search phrases flow correctly', () => {
        render(<PhrasesGridPage />);
        
        // Simular cambio en el search bar
        const searchInput = screen.getByTestId('search-bar-mock');
        fireEvent.change(searchInput, { target: { value: 'prueba 1' } });
        
        expect(mockUsePhrases.searchPhrases).toHaveBeenCalledWith('prueba 1');
    });
});