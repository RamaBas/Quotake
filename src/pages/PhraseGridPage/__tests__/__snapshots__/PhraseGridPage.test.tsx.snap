// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhrasesGridPage renders correctly with phrases 1`] = `
<DocumentFragment>
  <input
    data-testid="search-bar-mock"
    type="text"
  />
  <div
    data-testid="phrases-grid-mock"
  />
  <div
    data-open="false"
    data-testid="delete-modal-mock"
  />
</DocumentFragment>
`;
